### PR TITLE
Fix Bitbucket push event handling

### LINKED ISSUE #7683: BitBucket App: Cannot read properties of null (reading 'name')
This is an issue that we see for about a year now, but saw an steap increase after todays deployment.
GCloud Error: https://console.cloud.google.com/errors/CK203s2T2-64mwE?service=server&time=PT6H&project=gitpod-191109

### PR DESCRIPTION
**Pull request by AlexTugarev**
Fixes #7683

This PR also fixes the linking of prebuilds to the related project.

Screen Shot 2022-01-26 at 10 24 43

```release-note
Fix Bitbucket push event handling
```

### CODE REVIEW

**Comment by roboquat** [comment]
@AlexTugarev: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **alextugarev** after the PR has been reviewed.

Associated issue: *#7683*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by jankeromnes** [comment]
Taking a look now 👀

**Review by jankeromnes — COMMENTED** [review]
Many thanks for adding the project association for Bitbucket push events! 👍

This change looks good to me in principle, except for the part about impersonating a random team owner (please see inline).

I've also tested that triggering Prebuilds by pushing commits to Bitbucket still works 👍 (both for user projects and team projects).

File: components/server/ee/src/prebuilds/bitbucket-app.ts
Comment (AlexTugarev) [reply components/server/ee/src/prebuilds/bitbucket-app.ts:118]: > Why is it useful to pick a random owner from the team here, as opposed to just picking the user that started the prebuild as before?

Another explainer: have a user A create a project in a team and then let user A leave the team, which account to use to tag the prebuild?

Comment (AlexTugarev) [reply components/server/ee/src/prebuilds/bitbucket-app.ts:118]: @jankeromnes, can we discuss this in a call?

Comment (geropl) [reply components/server/ee/src/prebuilds/bitbucket-app.ts:118]: Hm. I tend to agree with Jan that we should not try to do sth intimate (use token) that is unexpected.
How does the BitBucket "installation flow" look like? E.g., can we store a userId there that we can use here? That way we'd at least maintain the expectation that who registered a repo will also execute the prebuilds (like for GitHub apps).

> Another explainer: have a user A create a project in a team and then let user A leave the team, which account to use to tag the prebuild?

This sounds like a very interesting question to me: What is `projectId.userId` useful for, anyway? :thinking:

File: components/server/ee/src/prebuilds/bitbucket-app.ts
Comment (geropl) [inline components/server/ee/src/prebuilds/bitbucket-app.ts:175]: :bulb:

File: components/server/ee/src/prebuilds/bitbucket-app.ts
Comment (geropl) [reply components/server/ee/src/prebuilds/bitbucket-app.ts:118]: **Suggestion**: For the sake of this PR, let's separate the "who owns the Team" and "who owns the BitBucket 'installation'" discussions into (a) separate issue(s). And simply exit here if `project.userId` is not set. Would that work, @AlexTugarev ?

Comment (AlexTugarev) [reply components/server/ee/src/prebuilds/bitbucket-app.ts:118]: The point is that we're storing the users token in the webhook registration, which is part of history when we had no project introduced yet. I believe it would be more robust to add the project id in the webhook registration and use any available token of any team owner (with access to that BB repo) when doing updates to the webook registration. This is something we concluded while discussing about the ownership of a project within a team, and what should happen when one leaves a team, e.g. should the project be removed? 

I'm also ok with trying to maintain the installer's link as far as possible and fallback to a team owner after the installer was removed from the team.

Comment (AlexTugarev) [reply components/server/ee/src/prebuilds/bitbucket-app.ts]: @jankeromnes and @geropl, I changed to prefer the previous webhook installer, which should meet the expectation in most cases. I suggest to overhaul the semantics for all providers in a separate issues, because it seems that we did not cover all cases. That's something we can and should cover in unit tests for clear communications.

File: components/server/ee/src/prebuilds/bitbucket-app.ts
Comment (geropl) [inline components/server/ee/src/prebuilds/bitbucket-app.ts:113]: @AlexTugarev IMO we need a comment here to explain why we need to find the project owner, and how we get there (project.userId, teamId, members with "owner" role).

File: components/server/ee/src/prebuilds/bitbucket-app.ts
Comment (geropl) [reply components/server/ee/src/prebuilds/bitbucket-app.ts]: Sounds good!

**Comment by geropl** [comment]
/hold
/lgtm

The code LGTM except that I think we should add a comment with the gist of our discussion to `findProjectOwner`.

Please ping for new /lgtm then!

File: components/server/ee/src/prebuilds/bitbucket-app.ts
Comment (AlexTugarev) [reply components/server/ee/src/prebuilds/bitbucket-app.ts]: @jankeromnes and @geropl, thanks once again for pushing this! after re-re-reading the flow and testing once more, I realized that's a bad decision to limit to project owners 🙄. the fallback mechanism should, if applied, try to find and team member to rescue the state of the project, if the original installer was removed from the team. so, priority is to maintain the webhook membership and then to fallback to any of the team member with the relevant provider connection.

I took the change and aligned the gitlab app code as well, because that was the blueprint here. 

let's continue the clean up of this in the effort to provide a state of the webhook installation.

File: components/server/ee/src/prebuilds/bitbucket-app.ts
Comment (geropl) [inline components/server/ee/src/prebuilds/bitbucket-app.ts:117]: :heart:

File: components/server/ee/src/prebuilds/gitlab-app.ts
Comment (geropl) [inline components/server/ee/src/prebuilds/gitlab-app.ts:153]: :bangbang: copy-paste error, should be: `Public-GitLab` @AlexTugarev

Comment (AlexTugarev) [reply components/server/ee/src/prebuilds/gitlab-app.ts:153]: on it

Comment (AlexTugarev) [reply components/server/ee/src/prebuilds/gitlab-app.ts]: done

**Review by geropl — APPROVED** [review]
Code looks good! :+1: 

Thx for pushing this till the end!

**Pull request merged**
